src/l500/l500-options.cpp

```diff
@@ -255,10 +349,63 @@ namespace librealsense
         return res;
     }
 
+    rs2_l500_visual_preset l500_options::calc_preset_from_controls()
+    {
+        rs2_l500_visual_preset preset = RS2_L500_VISUAL_PRESET_CUSTOM;
+        try
+        {
+            // compare default values to current values exept from laser power that can
+            // get diffrant value according to preset
+            std::map< rs2_option, float > hw_options_default_values;
+            for( auto control : _hw_options )
+            {
+                if( control.first != RS2_OPTION_LASER_POWER )
+                {
+                    if( control.second->get_range().def != control.second->query() )
+                        return preset;
+                }
+            }
+
+            // all the hw_options values are equal to their default values
+            // now what is left to check if the gain and laser power correspond to one of the
+            // presets
+            auto max_laser = _hw_options[RS2_OPTION_LASER_POWER]->get_range().max;
+            auto def_laser = _hw_options[RS2_OPTION_LASER_POWER]->get_range().def;
+
+            std::map< std::pair< rs2_digital_gain, float >, rs2_l500_visual_preset >
+                gain_and_laser_to_preset = {
+                    { { RS2_DIGITAL_GAIN_HIGH, def_laser }, RS2_L500_VISUAL_PRESET_NO_AMBIENT },
+                    { { RS2_DIGITAL_GAIN_LOW, max_laser }, RS2_L500_VISUAL_PRESET_LOW_AMBIENT },
+                    { { RS2_DIGITAL_GAIN_HIGH, max_laser }, RS2_L500_VISUAL_PRESET_MAX_RANGE },
+                    { { RS2_DIGITAL_GAIN_LOW, def_laser }, RS2_L500_VISUAL_PRESET_SHORT_RANGE },
+                };
+
+            if( gain_and_laser_to_preset.find( { ( rs2_digital_gain )(int)_digital_gain->query(),
+                                                 _hw_options[RS2_OPTION_LASER_POWER]->query() } )
+                != gain_and_laser_to_preset.end() )
+            {
+                return gain_and_laser_to_preset[{ ( rs2_digital_gain )(int)_digital_gain->query(),
+                                                  _hw_options[RS2_OPTION_LASER_POWER]->query() }];
+            }
+            return preset;
+        }
+
+        catch( ... )
+        {
+
+            LOG_ERROR( "Exception caught in calc_preset_from_controls" );
+            return preset;
+        }
+    }
     void l500_options::on_set_option(rs2_option opt, float value)
     {
         if (opt == RS2_OPTION_VISUAL_PRESET)
         {
+            if( static_cast< rs2_l500_visual_preset >( int( value ) )
+                == RS2_L500_VISUAL_PRESET_DEFAULT )
+                throw invalid_value_exception(
```

Comment:
This is after the option value has already been set, no?
Maybe it's worth it to have the visual preset in its own class and override `set`?

Reply:
There's some interaction with `verify_max_usable_range_restrictions` which may throw exceptions so should happen *before* the option is actually set.
I can see using a pre- and post- notification... up to you, but I'd rather this code used polymorphism rather than a function that had to check for special types.